pkg/apis/serving/v1alpha1/inference_graph.go

```diff
@@ -0,0 +1,257 @@
+// Multiple regression trees are often combined using various averaging techniques.
+// e.g tagging models with segment identifiers and weights to be used for their combination in these ways.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: ensemble
+// spec:
+//   nodes:
+//     root:
+//       routerType: Sequence
+//       routes:
+//       - service: feast
+//       - nodeName: ensembleModel
+//         data: $response
+//     ensembleModel:
+//       routerType: Ensemble
+//       routes:
+//       - service: sklearn-model
+//       - service: xgboost-model
+// ```
+//
+// Scoring a case using a sequence, or chain of models allows the output of one model to be passed in as input to the subsequent models.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: model-chainer
+// spec:
+//   nodes:
+//     root:
+//       routerType: Sequence
+//       routes:
+//       - service: mymodel-s1
+//       - service: mymodel-s2
+//         data: $response
+//       - service: mymodel-s3
+//         data: $response
+// ```
+//
+// In the flow described below, the pre_processing node base64 encodes the image and passes it to two model nodes in the flow.
+// The encoded data is available to both these nodes for classification. The second node i.e. dog-breed-classification takes the
+// original input from the pre_processing node along-with the response from the cat-dog-classification node to do further classification
+// of the dog breed if required.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: dog-breed-classification
+// spec:
+//   nodes:
+//     root:
+//       routerType: Sequence
+//       routes:
+//       - service: cat-dog-classifier
+//       - nodeName: breed-classifier
+//         data: $request
+//     breed-classifier:
+//       routerType: Switch
+//       routes:
+//       - service: dog-breed-classifier
+//         condition: { .predictions.class == "dog" }
+//       - service: cat-breed-classifier
+//         condition: { .predictions.class == "cat" }
+// ```
+type InferenceRouter struct {
+	// RouterType
+	//
+	// - `Sequence:` chain multiple inference steps with input/output from previous step
+	//
+	// - `Splitter:` randomly routes to the target service according to the weight
+	//
+	// - `Ensemble:` routes the request to multiple models and then merge the responses
+	//
+	// - `Switch:` routes the request to one of the steps based on condition
+	//
+	RouterType InferenceRouterType `json:"routerType"`
+
+	// Steps defines destinations for the current router node
+	// +optional
+	Steps []InferenceStep `json:"steps,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+// Exactly one InferenceTarget field must be specified
+
+type InferenceTarget struct {
+	// The node name for routing as next step
+	// +optional
+	NodeName string `json:"nodeName,omitempty"`
+
+	// named reference for InferenceService
+	ServiceName string `json:"serviceName,omitempty"`
```

Comment:
I noticed that in the example InferenceGraph YAML files and doc snippets,  `service` is used instead of `serviceName` as declared here and in the generated CRD. Should YAML files and docs be updated with `serviceName`?

Reply:
updated